### PR TITLE
[Serializer] Give more hints when an attribute is not correctly used

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | kind of
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

before : 
image

after

image

### CODE REVIEW

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 1]
```diff
@@ -157,7 +157,21 @@ public function loadAnnotations(object $reflector): iterable
     {
         foreach ($reflector->getAttributes() as $attribute) {
             if ($this->isKnownAttribute($attribute->getName())) {
-                yield $attribute->newInstance();
+                try {
+                    yield $attribute->newInstance();
+                } catch (\Error $th) {
+                    $on = '';
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Maybe could you use a `match(true)` instead of these if?

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 2]
```diff
@@ -157,7 +157,21 @@ public function loadAnnotations(object $reflector): iterable
     {
         foreach ($reflector->getAttributes() as $attribute) {
             if ($this->isKnownAttribute($attribute->getName())) {
-                yield $attribute->newInstance();
+                try {
+                    yield $attribute->newInstance();
+                } catch (\Error $th) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: should be named `$e` I guess?

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 3]
```diff
@@ -157,7 +157,21 @@ public function loadAnnotations(object $reflector): iterable
...
+                    if ($reflector instanceof \ReflectionClass) {
+                        $on = ' on class '.$reflector->name;
+                    }
+                    if ($reflector instanceof \ReflectionMethod) {
+                        $on = sprintf(' on "%s::%s()"', $reflector->getDeclaringClass()->name, $reflector->name);
+                    }
+                    if ($reflector instanceof \ReflectionProperty) {
+                        $on = sprintf(' on "%s::$%s"', $reflector->getDeclaringClass()->name, $reflector->name);
+                    }
+                    throw new MappingException(sprintf('Could not instantiate attribute%s.', $on), 0, $th);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: shouldn't we add the name of the attribute in the message?

**Comment by lyrixx** [comment]
Thanks for the review, I addressed your comments

**Review by dunglas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
(please rebase to get rid of bad commits)

**Comment by lyrixx** [comment]
@nicolas-grekas rebase done 👍🏼

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 4]
```diff
@@ -157,7 +157,18 @@ public function loadAnnotations(object $reflector): iterable
     {
         foreach ($reflector->getAttributes() as $attribute) {
             if ($this->isKnownAttribute($attribute->getName())) {
-                yield $attribute->newInstance();
+                try {
+                    yield $attribute->newInstance();
+                } catch (\Error $e) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:162]: ~I'm always mixed on catching `Error` but I guess there is a specific type you want to catch?~

saw the screenshot in the PR description :)

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 5]
```diff
@@ -157,7 +157,18 @@ public function loadAnnotations(object $reflector): iterable
...
+                    $on = match (true) {
+                        $reflector instanceof \ReflectionClass => ' on class '.$reflector->name,
+                        $reflector instanceof \ReflectionMethod => sprintf(' on "%s::%s()"', $reflector->getDeclaringClass()->name, $reflector->name),
+                        $reflector instanceof \ReflectionProperty => sprintf(' on "%s::$%s"', $reflector->getDeclaringClass()->name, $reflector->name),
+                        default => '',
+                    };
+
+                    throw new MappingException(sprintf('Could not instantiate attribute "%s"%s.', $attribute->getName(), $on), 0, $e);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:173]: shouldn't we add $e->getMessage() at the end of the message?

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:162]: should we throw when `get_class($e) !== Error::class`?

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:173]: The new exception already wrap the previous one. You already made me add the attribute (which is redundant too). you can see the initial screenshot in the PR, all information are here to debug the situation I think

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 4 above)
Comment (lyrixx) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:162]: I don't like to catch such wide error too. But this is the best type I guess:

image

Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:162]: Can you add the check to throw when `get_class($e) !== \Error::class`?

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:173]: OK, works for me.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 4 above)
Comment (lyrixx) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:162]: done (with $e::class, hope it's not an issue ; I forgot to re-read your comment before patchin)

For the ref: https://heap.space/xref/PHP-8.1/ext/reflection/php_reflection.c?r=59dd4fd7#6596

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**